File: redirectglide/src/main/java/com/aminography/redirectglide/OkHttpApiCallUrlFetcher.java
package com.aminography.redirectglide;

import android.support.annotation.NonNull;

import com.bumptech.glide.load.HttpException;
import com.bumptech.glide.load.model.GlideUrl;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.Response;

/**
 * Fetches an {@link GlideUrl} using the okhttp library.
 */
public class OkHttpApiCallUrlFetcher extends OkHttpRedirectUrlFetcher implements okhttp3.Callback {

    public OkHttpApiCallUrlFetcher(Call.Factory client, BaseApiCallGlideUrl url) {
        super(client, url);
    }

    @Override
    public void consumeResponse(@NonNull Response response) {
        responseBody = response.body();
        if (response.isSuccessful() && responseBody != null) {
            try {
                String sourceUrl = ((BaseApiCallGlideUrl) url).extractImageUrl(responseBody.string());
                callback.onDataReady(new GlideUrl(sourceUrl));
            } catch (IOException e) {
                callback.onLoadFailed(new HttpException(response.message(), response.code()));
                e.printStackTrace();
            }
        } else {
            callback.onLoadFailed(new HttpException(response.message(), response.code()));
        }
    }

}

File: redirectglide/src/main/java/com/aminography/redirectglide/OkHttpApiCallUrlFetcher.kt
package com.aminography.redirectglide

import com.bumptech.glide.load.HttpException
import com.bumptech.glide.load.model.GlideUrl
import okhttp3.Call
import okhttp3.Callback
import okhttp3.Response
import java.io.IOException

/**
 * Fetches an [GlideUrl] using the okhttp library.
 */
class OkHttpApiCallUrlFetcher(
    client: Call.Factory,
    url: BaseApiCallGlideUrl
) : OkHttpRedirectUrlFetcher(client, url), Callback {

    override fun consumeResponse(response: Response) {
        responseBody = response.body
        if (response.isSuccessful && responseBody != null) {
            try {
                val sourceUrl = (url as BaseApiCallGlideUrl).extractImageUrl(responseBody?.string())
                callback?.onDataReady(GlideUrl(sourceUrl))
            } catch (e: IOException) {
                callback?.onLoadFailed(HttpException(response.message, response.code))
                e.printStackTrace()
            }
        } else {
            callback?.onLoadFailed(HttpException(response.message, response.code))
        }
    }

}